src/main/java/org/opentripplanner/graph_builder/module/interlining/InterlineProcessor.java

```diff
@@ -0,0 +1,184 @@
+
+    LOG.info("Finding interlining trips based on block IDs.");
+    for (TripPattern pattern : tripPatterns) {
+      Timetable timetable = pattern.getScheduledTimetable();
+      /* TODO: Block semantics seem undefined for frequency trips, so skip them? */
+      for (TripTimes tripTimes : timetable.getTripTimes()) {
+        Trip trip = tripTimes.getTrip();
+        if (!Strings.isNullOrEmpty(trip.getGtfsBlockId())) {
+          tripTimesForBlock.put(BlockIdAndServiceId.ofTrip(trip), tripTimes);
+          // For space efficiency, only record times that are part of a block.
+          patternForTripTimes.put(tripTimes, pattern);
+        }
+      }
+    }
+
+    // Associate pairs of TripPatterns with lists of trips that continue from one pattern to the other.
+    Multimap<P2<TripPattern>, P2<Trip>> interlines = ArrayListMultimap.create();
+
+    // Sort trips within each block by first departure time, then iterate over trips in this block and service,
+    // linking them. Has no effect on single-trip blocks.
+    SERVICE_BLOCK:for (BlockIdAndServiceId block : tripTimesForBlock.keySet()) {
+      List<TripTimes> blockTripTimes = tripTimesForBlock.get(block);
+      Collections.sort(blockTripTimes);
+      TripTimes prev = null;
+      for (TripTimes curr : blockTripTimes) {
+        if (prev != null) {
+          if (prev.getDepartureTime(prev.getNumStops() - 1) > curr.getArrivalTime(0)) {
+            LOG.error(
+              "Trip times within block {} are not increasing on service {} after trip {}.",
+              block.blockId(),
+              block.serviceId(),
+              prev.getTrip().getId()
+            );
+            continue SERVICE_BLOCK;
+          }
+          TripPattern prevPattern = patternForTripTimes.get(prev);
+          TripPattern currPattern = patternForTripTimes.get(curr);
+          var fromStop = prevPattern.lastStop();
+          var toStop = currPattern.firstStop();
+          double teleportationDistance = SphericalDistanceLibrary.fastDistance(
+            fromStop.getLat(),
+            fromStop.getLon(),
+            toStop.getLat(),
+            toStop.getLon()
+          );
+          if (teleportationDistance > maxInterlineDistance) {
+            issueStore.add(
+              new InterliningTeleport(prev.getTrip(), block.blockId(), (int) teleportationDistance)
+            );
+            // Only skip this particular interline edge; there may be other valid ones in the block.
+          } else {
+            interlines.put(
+              new P2<>(prevPattern, currPattern),
+              new P2<>(prev.getTrip(), curr.getTrip())
+            );
+          }
+        }
+        prev = curr;
+      }
+    }
+
+    return interlines;
+  }
+}
+
+/**
+ * This compound key object is used when grouping interlining trips together by (serviceId,
+ * blockId).
+ */
+record BlockIdAndServiceId(String blockId, FeedScopedId serviceId) {
```

Comment:
```suggestion
private record BlockIdAndServiceId(String blockId, FeedScopedId serviceId) {
```

Reply:
I tried this but the java compiler says that it's not allowed. Would you like me to move it into the class?